libraries/AP_VideoTX/AP_Tramp.cpp

```diff
@@ -0,0 +1,487 @@
+            device_limits.rf_freq_max = response_buffer[4]|(response_buffer[5] << 8);
+            device_limits.rf_power_max = response_buffer[6]|(response_buffer[7] << 8);
+            debug("device limits: min freq: %u, max freq: %u, max power %u",
+                unsigned(device_limits.rf_freq_min), unsigned(device_limits.rf_freq_max), unsigned(device_limits.rf_power_max));
+            return 'r';
+        }
+        break;
+    }
+    case 'v': {
+        const uint16_t freq = response_buffer[2]|(response_buffer[3] << 8);
+        // Check we're not reading the request (indicated by freq zero)
+        if (freq != 0) {
+            // Got response, update device status
+            const uint16_t power = response_buffer[4]|(response_buffer[5] << 8);
+            cur_control_mode = response_buffer[6]; // Currently only used for race lock
+            const bool pit_mode = response_buffer[7];
+            cur_act_power = response_buffer[8]|(response_buffer[9] << 8);
+
+            // update the vtx
+            AP_VideoTX& vtx = AP::vtx();
+            vtx.set_frequency_mhz(freq);
+
+            AP_VideoTX::VideoBand band;
+            uint8_t channel;
+            if (vtx.get_band_and_channel(freq, band, channel)) {
+                vtx.set_band(band);
+                vtx.set_channel(channel);
+            }
+
+            vtx.set_power_mw(power);
+            if (pit_mode) {
+                vtx.set_options(vtx.get_options() | uint8_t(AP_VideoTX::VideoOptions::VTX_PITMODE));
+            } else {
+                vtx.set_options(vtx.get_options() & ~uint8_t(AP_VideoTX::VideoOptions::VTX_PITMODE));
+            }
+
+            // make sure the configured values now reflect reality
+            vtx.set_defaults();
+
+            debug("device config: freq: %u, power: %u, pitmode: %u",
+                unsigned(freq), unsigned(power), unsigned(pit_mode));
+
+
+            return 'v';
+        }
+        break;
+    }
+    case 's': {
+        const uint16_t temp = (int16_t)(response_buffer[6]|(response_buffer[7] << 8));
+        // Check we're not reading the request (indicated by temp zero)
+        if (temp != 0) {
+            // Got response, update device status
+            cur_temp = temp;
+            return 's';
+        }
+        break;
+    }
+    }
+
+    // Likely reading a request, return zero to indicate not accepted
+    return 0;
+}
+
+// Reset receiver state machine
+void AP_Tramp::reset_receiver(void)
+{
+    receive_state = ReceiveState::S_WAIT_LEN;
+    receive_pos = 0;
+}
+
+// returns completed response code or 0
+char AP_Tramp::receive_response()
+{
+    if (port == nullptr) {
+        return 0;
+    }
+
+    // wait for complete packet
+    const uint16_t bytesNeeded = TRAMP_BUF_SIZE - receive_pos;
+    if (port->available() < bytesNeeded) {
+        return 0;
+    }
+
+    for (uint16_t i = 0; i < bytesNeeded; i++) {
+        const uint8_t c = port->read();
+        response_buffer[receive_pos++] = c;
```

Comment:
can receive_pos end up equal to TRAMP_BUF_SIZE and not be in the right state? it would get bytesNeeded==0
I think we need a specific check for bytesNeeded=0 to check state and reset